File: 0046-permutations/0046-permutations.java
class Solution {
    private void permutations(int ind,int[] nums,List<List<Integer>> ans){
        if(ind==nums.length){
            List<Integer> ds = new ArrayList<>();
            for(int i =0;i<nums.length;i++){
                ds.add(nums[i]);
            }
            ans.add(ds);
            return;
        }
        for(int i =ind;i<nums.length;i++){
            swap(i,ind,nums);
            permutations(ind+1,nums,ans);
            swap(i,ind,nums);
        }
    }
    private void swap(int i,int ind,int[] nums){
        int t = nums[i];
        nums[i] =nums[ind];
        nums[ind] = t;
    }
    public List<List<Integer>> permute(int[] nums) {
//         Optimized space complexity
        List<List<Integer>> ans = new ArrayList<>();
        permutations(0,nums,ans);
        return ans;
    }
}

File: 0046-permutations/0046-permutations.cpp
class Solution {
private:
    void f(int ind,vector<int> &nums,vector<int> ds,vector<vector<int>> &ans,vector<int> &freq) {
        if(ds.size() == nums.size()) {
            ans.push_back(ds);
            return;
        }
        for(int i = 0; i < nums.size(); i++) {
            if(!freq[i]) {
                ds.push_back(nums[i]);
                freq[i] = 1;
                f(i+1,nums,ds,ans,freq);
                ds.pop_back();
                freq[i] = 0;
            }
        }
    }
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> ans;
        vector<int> ds;
        vector<int> freq(nums.size(),0);
        f(0,nums,ds,ans,freq);
        return ans;
    }
};